### PR TITLE
ref(server): Move body compression into EnvelopeManager (INGEST-357)

### PR DESCRIPTION
**Pull request by Michcioperz**
While researching performance of Relay I discovered that the body compression for upstream requests is run on the system arbiter instead of on a thread pool, and that removing compression increases throughput drastically.

This change moves the workload of compression from UpstreamRelay to EnvelopeProcessor, which runs on a SyncArbiter.

### CODE REVIEW

File: relay-server/src/actors/envelopes.rs [hunk 1]
```diff
@@ -2002,11 +1997,23 @@ impl EnvelopeManager {
         }
 
         relay_log::trace!("sending envelope to sentry endpoint");
+
+        let http_encoding = self.config.http_encoding();
+        let (envelope_body, envelope_meta) = match Self::encode_envelope(envelope, http_encoding) {
+            Ok(v) => v,
+            Err(e) => {
+                return Box::new(future::err(SendEnvelopeError::SendFailed(
```
Comment (Michcioperz) [inline relay-server/src/actors/envelopes.rs]: i'm not particularly happy about the type of this error cause it's all backwards. i was thinking of making a new variant of SendEnvelopeError (two actually, a BodyEncodingError(std::io::Error) and an EnvelopeBuildError(EnvelopeError)), but then i'd have to carry them into ProcessingError too, and i'm not 100% sure it's worth it? idk

Comment (untitaker) [reply relay-server/src/actors/envelopes.rs]: I'd do exactly that, it's fine to have so many variants

Comment (Michcioperz) [reply relay-server/src/actors/envelopes.rs]: ok will do

**Comment by Michcioperz** [comment]
i ran some preliminary benchmarks and i get 1500 rps? not very cool.
i think actually envelopemanager doesn't have a syncarbiter?

File: relay-server/src/actors/envelopes.rs [hunk 2]
```diff
@@ -121,8 +125,14 @@ enum ProcessingError {
     #[fail(display = "could not serialize event payload")]
     SerializeFailed(#[cause] serde_json::Error),
 
+    #[fail(display = "could not build envelope for upstream")]
+    EnvelopeBuildFailed(#[cause] EnvelopeError),
+
+    #[fail(display = "could not encode request body")]
+    BodyEncodingFailed(#[cause] std::io::Error),
```
Comment (untitaker) [inline relay-server/src/actors/envelopes.rs:132]: would map all of this to None for now, I think this change makes sense but if we want to do that I would reevaluate Self::UpstreamRequestFailed too

File: relay-server/src/actors/envelopes.rs [hunk 3]
```diff
@@ -2016,10 +2029,55 @@ impl EnvelopeManager {
...
+
+    fn encode_envelope(
+        mut envelope: Envelope,
+        http_encoding: HttpEncoding,
+    ) -> Result<(Vec<u8>, RequestMeta), SendEnvelopeError> {
+        // Override the `sent_at` timestamp. Since the envelope went through basic
+        // normalization, all timestamps have been corrected. We propagate the new
+        // `sent_at` to allow the next Relay to double-check this timestamp and
+        // potentially apply correction again. This is done as close to sending as
+        // possible so that we avoid internal delays.
```
Comment (untitaker) [inline relay-server/src/actors/envelopes.rs:2117]: Note: I think we need to live with this. This value will be horribly wrong if there is significant amount of data piling up in the queue but if we need to fix that, I question why double-checking is done at all.

**Review by untitaker — CHANGES_REQUESTED** [review]
after the singular change to error mapping this is good to merge without another review

**Review by untitaker — CHANGES_REQUESTED** [review]
>i ran some preliminary benchmarks and i get 1500 rps? not very cool. i think actually envelopemanager doesn't have a syncarbiter?

this is actually a blocker... discussing how to solve this

File: relay-server/src/actors/envelopes.rs (hunk 3 above)
Comment (Michcioperz) [reply relay-server/src/actors/envelopes.rs:2117]: y'know this comment is not mine, i'm just copying it along with the rest

Comment (untitaker) [reply relay-server/src/actors/envelopes.rs:2117]: That's right but you were moving the code below this comment from after the upstream actor queue to before it

Comment (Michcioperz) [reply relay-server/src/actors/envelopes.rs:2117]: hmm tbh i think i don't quite understand what double checking we have

File: relay-server/src/actors/envelopes.rs (hunk 2 above)
Comment (Michcioperz) [reply relay-server/src/actors/envelopes.rs:132]: done in 8fca0ef

**Comment by Michcioperz** [comment]
benchmarking again against fake-sentry+nginx. 1000 clients over 4 locust workers. relay pinned to 6 cpu cores. just reading the rps gauge, so no flamegraph overhead.

- before change, identity: 3500+ rps
- before change, gzip: ~1500 rps
- after change, identity: ~3400 rps
- after change, gzip: ~2400 rps

so gzip is spectacularly good here imo but i'm worried about identity. thinking either a small refactor to skip encoding handling for identity or a canary to see if it matters?

**Review by untitaker — APPROVED** [review]
approving for canary

**Review by untitaker — APPROVED** [review]
let' ship

**Pull request merged**